src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -0,0 +1,145 @@
+            Requires.NotNull(activeConfiguredProjectSubscriptionService, nameof(activeConfiguredProjectSubscriptionService));
+
+            _commonServices = commonServices;
+            _contextProvider = contextProvider;
+            _tasksService = tasksService;
+            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+
+            Handlers = new OrderPrecedenceImportCollection<ILanguageServiceRuleHandler>(projectCapabilityCheckProvider: commonServices.Project);
+        }
+
+        [ImportMany]
+        public OrderPrecedenceImportCollection<ILanguageServiceRuleHandler> Handlers
+        {
+            get;
+        }
+
+        [ProjectAutoLoad(ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService2)]
+        private Task OnProjectFactoryCompletedAsync()
+        {
+            using (_tasksService.LoadedProject())
+            {
+                var watchedEvaluationRules = GetWatchedRules(RuleHandlerType.Evaluation);
+                var watchedDesignTimeBuildRules = GetWatchedRules(RuleHandlerType.DesignTimeBuild);
+
+                _designTimeBuildSubscriptionLink = _activeConfiguredProjectSubscriptionService.JointRuleSource.SourceBlock.LinkTo(
+                  new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(e => OnProjectChanged(e, RuleHandlerType.DesignTimeBuild)),
+                  ruleNames: watchedDesignTimeBuildRules.Union(watchedEvaluationRules), suppressVersionOnlyUpdates: true);
+
+                _evaluationSubscriptionLink = _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock.LinkTo(
+                    new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(e => OnProjectChanged(e, RuleHandlerType.Evaluation)),
+                    ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates: true);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        protected async override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Don't initialize if we're unloading
+            _tasksService.UnloadCancellationToken.ThrowIfCancellationRequested();
+
+            _projectContext = await _contextProvider.Value.CreateProjectContextAsync()
+                                                          .ConfigureAwait(false);
+
+            foreach (var handler in Handlers)
+            {
+                handler.Value.SetContext(_projectContext);
+            }
+        }
+
+        private async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> e, RuleHandlerType handlerType)
+        {
+            if (IsDisposing || IsDisposed)
+                return;
+
+            await InitializeAsync().ConfigureAwait(false);
+
+            // TODO: https://github.com/dotnet/roslyn-project-system/issues/353
+            await _commonServices.ThreadingService.SwitchToUIThread();
```

Comment:
I presume this ensures Handle is invoked from a single thread for now. Eventually, when we remove the switch to UI thread, it is likely we invoke Handle from multiple threads for project changes? We will need to ensure that two project change events don't stomp over each other:

1. If we can receive project changed events in different order, then we need a versioning scheme to ensure that we don't corrupt the state - a subsequent project change might be dependent on a prior project change.
2. We will need to put locking in place here or in the handlers to ensure each project change is processed atomically.